ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/Node/GSM.hs

```diff
@@ -0,0 +1,465 @@
+                getDurationUntilTooOld wd sno >>= \case
+                    After{}     -> return YoungEnough
+                    Already     -> do
+                        removeMarkerFile markerFileView
+                        return TooOld
+
+{-------------------------------------------------------------------------------
+  A real implementation
+-------------------------------------------------------------------------------}
+
+-- | The actual GSM logic for boot strap peers
+--
+-- See the @BootstrapPeersIER.md@ document for the specification of this logic.
+realGsmEntryPoints :: forall m upstreamPeer selection candidate.
+     ( SI.MonadDelay m
+     , SI.MonadTimer m
+     , Eq upstreamPeer
+     )
+  => GsmView m upstreamPeer selection candidate
+  -> GsmEntryPoints m
+realGsmEntryPoints gsmView = GsmEntryPoints {
+    enterCaughtUp
+  ,
+    enterOnlyBootstrap
+  }
+  where
+    GsmView {
+        antiThunderingHerd
+      ,
+        candidateOverSelection
+      ,
+        durationUntilTooOld
+      ,
+        equivalent
+      ,
+        getChainSyncCandidates
+      ,
+        getChainSyncIdlers
+      ,
+        getCurrentSelection
+      ,
+        minCaughtUpDuration
+      ,
+        setCaughtUpPersistentMark
+      ,
+        writeLedgerStateJudgement
+      } = gsmView
+
+    enterCaughtUp :: forall neverTerminates. m neverTerminates
+    enterCaughtUp  = enterCaughtUp' antiThunderingHerd
+
+    enterOnlyBootstrap :: forall neverTerminates. m neverTerminates
+    enterOnlyBootstrap  = enterOnlyBootstrap' antiThunderingHerd
+
+    enterCaughtUp' :: forall neverTerminates. Maybe StdGen -> m neverTerminates
+    enterCaughtUp' g = do
+        g' <- blockWhileCaughtUp g
+
+        setCaughtUpPersistentMark False
+        writeLedgerStateJudgement TooOld
+
+        enterOnlyBootstrap' g'
+
+    enterOnlyBootstrap' :: Maybe StdGen -> forall neverTerminates. m neverTerminates
+    enterOnlyBootstrap' g = do
+        blockUntilCaughtUp
+
+        writeLedgerStateJudgement YoungEnough
+        setCaughtUpPersistentMark True
+
+        -- When transitioning from OnlyBootstrap to CaughtUp, the node will
+        -- remain in CaughtUp for at least 'minCaughtUpDuration', regardless of
+        -- the selection's age.
+        SI.threadDelay $ realToFrac minCaughtUpDuration
+
+        enterCaughtUp' g
+
+    blockWhileCaughtUp :: Maybe StdGen -> m (Maybe StdGen)
+    blockWhileCaughtUp g = do
+        -- Randomly add up to 5min.
+        --
+        -- Under the ideal circumstances, nodes have perfectly synchronized
+        -- clocks. However, if there's a block production outage, that means
+        -- /all/ nodes will switch back to the bootstrap peers
+        -- /simultaneously/, incurring a thundering herd of requests on that
+        -- relatively small population. This random change will spread that
+        -- load out.
+        --
+        -- TODO should the Diffusion Layer do this? IE the node /promptly/
```

Comment:
A question for the Networking team. @bolt12 @coot @njd42

In this commit, the GSM will (excluding some exceptional scenarios) transition from CaughtUp (aka `YoungEnough`) back to OnlyBootstrap (aka `TooOld`) once its selection exceeds some age that is randomly chosen between 20min and 25min.

- The randomness is to avoid thundering herds "attacking" the bootstrap peers whenever `mainnet` block production does lapse for >20min.

- The IER/our discussions indicate that the GSM should remain in CaughtUp until the selection is at least 20min old --- we don't want the network to "panic" too frequently, and 20min is what PNSol recommended as "robust enough".

But it occurs to me now: when the selection more than 20min old, we should not simply stay in CaughtUp. Any time beyond that is extra exposure while vulnerable (eg perhaps we're eclipsed). So, how about the following subtly different behavior instead?

- After 20min, the GSM _promptly_ transitions from CaughtUp to OnlyBootstrap.
- As per usual, this promptly causes the Diffusion layer to disconnect from all current peers. (Aside: it could keep any that happen to be bootstrap peers.)
- But then, instead of immediately connecting to bootstrap peers, the Diffusion layer intentionally leaves the valency at 0 for a random duration between 0 and 5min.

The key benefit is that: the node is safe while the GSM is in OnlyBootstrap and the valency is 0.

The key cost is that: when there actually is a very long block production outage, then the randomness would manifest as an expectation of 11% of (honest) nodes intentionally having 0 valency at any given time. E(Uniform(0,5)) = 2.5min, and 2.5min / (20min+2.5min) is 10/9 ~ 11%.

Reply:
You mention the thundering herd problem for a particular edge case where there's a global outage and want a random delay to not over stress the bootstrap peers, and if I understood you correctly you think it doesn't make sense to switch to `TooOld` later due to the possibility of being eclipsed. I say it the eclipse part doesn't matter too much since the outage is global, right ?

But considering your proposal, who's to say that 0-5 min is enough for the other nodes to connect and become caught up? Should this delay be decided based on a statistical likelihood of the time it would take for a node to catch up after a 20+ min outage?

Reply:
> I say it the eclipse part doesn't matter too much since the outage is global, right ?

The mechanism exists because of the global possibility, but the mechanism needs to behave well even if it's not global, since the individual node unfortunately cannot distinguish the two scenarios.

Reply:
> Should this delay be decided based on a statistical likelihood of the time it would take for a node to catch up after a 20+ min outage?

Yes. I was roughly ballparking that around 5min: 20 min should include around 60 active slots, and I anticipate a node should sync ~60 blocks from the bootstrap peers in much less than 5 minutes. (Of course, these numbers might change if we change the Cardano block dynamics, etc.)

Reply:
For the main question, though: what do you think about moving the "anti-thundering herd" delay into the Diffusion Layer, instead of baking it into the Consensus Layer's state machine?

Reply:
from what I could understand the "anti-thundering herd" solution seems a lot simpler if implemented on consensus side. On networking side it would require significantly more work, so with this being said I'd say I think it should leave on consensus. I agree that ideally it should be implemented on the networking side but for a temporary feature I believe we should keep it as simple as possible.